README.md
A minimalistic framework-agnostic library for showing info messages without
interrupting the overall flow on web applications.

## Usage

A simple usage example:

```dart
import 'package:toast/toast.dart';

main() {
  Toast(
    title: 'Info',
    text: 'Your changes has been saved!',
    duration: const Duration(seconds: 3));
}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/afpatmin/toast/issues


CHANGELOG.md
## 1.0.4-nullsafety
- Fixed issue where elements to the left and right toast couldn't be clicked

## 1.0.3-nullsafety
- Fixed issue where body line-height affected close icon placement
- Fixed issue where the Toast didn't fade out when removed
- Toast font is now set to the same font as the document body

## 1.0.2-nullsafety
- Header and text sizes are now enforced to 1.2rem/1rem

## 1.0.1-nullsafety
- Made the Toasts responive using flex
- Set default font sizes for title (1.2rem) and text (1rem)

## 1.0.0-nullsafety
- Initial version


test/toast_test.dart
import 'package:toast/toast.dart';
import 'package:test/test.dart';

void main() {}


lib/src/alert_level_properties.dart
enum AlertLevel { success, info, warning, error }

class AlertLevelProperties {
  final String backgroundColor;
  final String color;
  final String iconSVG;
  final String className;

  const AlertLevelProperties({
    required this.backgroundColor,
    required this.color,
    required this.iconSVG,
    required this.className,
  });
}

const Map<AlertLevel, AlertLevelProperties> properties = {
  AlertLevel.error: AlertLevelProperties(
    backgroundColor: '#E74851',
    color: '#fff',
    iconSVG: '''
        <svg xmlns="http://www.w3.org/2000/svg" height="24" viewBox="0 0 24 24" width="24">
          <path d="M0 0h24v24H0z" fill="none"/>
          <path fill="#fff" d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"/>
        </svg>''',
    className: 'toast-error',
  ),
  AlertLevel.info: AlertLevelProperties(
    backgroundColor: '#264653',
    color: '#fff',
    iconSVG: '''
          <svg xmlns="http://www.w3.org/2000/svg" height="24" viewBox="0 0 24 24" width="24">
            <path d="M0 0h24v24H0z" fill="none" />
            <path fill="#fff" d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"/>
          </svg>''',
    className: 'toast-info',
  ),
  AlertLevel.success: AlertLevelProperties(
    backgroundColor: '#2A9D8F',
    color: '#fff',
    iconSVG: '''
        <svg xmlns="http://www.w3.org/2000/svg" height="24" viewBox="0 0 24 24" width="24">
          <path d="M0 0h24v24H0z" fill="none"/>
          <path fill="#fff" d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"/>
        </svg>''',
    className: 'toast-success',
  ),
  AlertLevel.warning: AlertLevelProperties(
    backgroundColor: '#f87F62',
    color: '#fff',
    iconSVG: '''
        <svg xmlns="http://www.w3.org/2000/svg" height="24" viewBox="0 0 24 24" width="24">
          <path d="M0 0h24v24H0z" fill="none"/>
          <circle cx="12" cy="19" r="2" fill="#fff"/>
          <path d="M10 3h4v12h-4z" fill="#fff"/>
        </svg>''',
    className: 'toast-warning',
  ),
};


lib/src/toast.dart
import 'dart:async';
import 'dart:html' as dom;
import 'dart:svg';

import 'alert_level_properties.dart';

final int _toastMargin = 4;

class Toast {
  final AlertLevel alertLevel;
  final dom.DivElement _container = dom.DivElement();
  final dom.DivElement _content = dom.DivElement();
  final dom.DivElement _closeButton = dom.DivElement()
    ..className = 'af-button-close'
    ..style.cursor = 'pointer'
    ..style.fontSize = '0'
    ..append(SvgElement.svg('''
        <svg xmlns="http://www.w3.org/2000/svg" height="24" viewBox="0 0 24 24" width="24">
          <path d="M0 0h24v24H0z" fill="none"/>
          <path fill="#fff" d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/>
        </svg>
        '''));
  final SvgElement _icon;
  final dom.HeadingElement _heading;
  final dom.ParagraphElement _text;
  Timer? timer;

  int get toastCount =>
      dom.document.body?.querySelectorAll('div.af-toast-container').length ?? 0;

  factory Toast.error(
          {String title = 'Error',
          required String text,
          int width = 600,
          int padding = 12,
          int iconSize = 24,
          Duration? duration}) =>
      Toast._(
          title: title,
          text: text,
          width: width,
          padding: padding,
          iconSize: iconSize,
          duration: duration,
          alertLevel: AlertLevel.error);

  factory Toast.info(
          {String title = 'Information',
          required String text,
          int width = 600,
          int padding = 12,
          int iconSize = 24,
          Duration? duration}) =>
      Toast._(
          title: title,
          text: text,
          width: width,
          padding: padding,
          iconSize: iconSize,
          duration: duration,
          alertLevel: AlertLevel.info);

  factory Toast.success(
          {String title = 'Success',
          required String text,
          int width = 600,
          int padding = 12,
          int iconSize = 24,
          Duration? duration}) =>
      Toast._(
          title: title,
          text: text,
          width: width,
          padding: padding,
          iconSize: iconSize,
          duration: duration,
          alertLevel: AlertLevel.success);

  factory Toast.warning(
          {String title = 'Warning',
          required String text,
          int width = 600,
          int padding = 12,
          int iconSize = 24,
          Duration? duration}) =>
      Toast._(
          title: title,
          text: text,
          width: width,
          padding: padding,
          iconSize: iconSize,
          duration: duration,
          alertLevel: AlertLevel.warning);

  Toast._({
    this.alertLevel = AlertLevel.info,
    required String title,
    required String text,
    int width = 600,
    int padding = 12,
    int iconSize = 24,
    Duration? duration,
  })  : _heading = dom.HeadingElement.h2()
          ..className = 'af-heading'
          ..style.fontSize = '1.2rem'
          ..style.fontWeight = '600'
          ..style.margin = '0'
          ..style.overflow = 'hidden'
          ..style.textOverflow = 'ellipsis'
          ..style.whiteSpace = 'nowrap'
          ..innerText = title,
        _text = dom.ParagraphElement()
          ..className = 'af-text'
          ..styleMap?.set('all', 'initial')
          ..style.fontSize = '1rem'
          ..style.lineHeight = '1rem'
          ..style.margin = '0 0 0 8px'
          ..style.flex = '1'
          ..innerText = text,
        _icon = SvgElement.svg(properties[alertLevel]!.iconSVG)
          ..attributes['width'] = '$iconSize'
          ..attributes['height'] = '$iconSize' {
    final property = properties[alertLevel]!;
    _container
      ..classes = ['af-toast-container']
      ..style.pointerEvents = 'none'
      ..style.position = 'fixed'
      ..style.zIndex = '${999 - toastCount}'
      ..style.bottom = '-100px'
      ..style.left = '0'
      ..style.width = '100vw'
      ..style.display = 'flex'
      ..style.justifyContent = 'center'
      ..style.transition = 'bottom 200ms ease, opacity 300ms ease';

    _content
      ..classes = ['af-toast', property.className]
      ..style.borderRadius = '4px'
      ..style.padding = '${padding}px'
      ..style.marginLeft = '${_toastMargin}px'
      ..style.marginRight = '${_toastMargin}px'
      ..style.width = '${width}px'
      ..style.backgroundColor = property.backgroundColor;

    final font = dom.document.body?.getComputedStyle().fontFamily;

    _heading.style.color = property.color;
    _closeButton.style.color = property.color;
    _text.style.color = property.color;
    _heading.style.fontFamily = font;
    _text.style.fontFamily = font;

    final titleRow = dom.DivElement()
      ..className = 'af-title'
      ..style.display = 'flex'
      ..style.alignItems = 'center'
      ..style.justifyContent = 'space-between'
      ..append(_heading)
      ..append(_closeButton);

    final textRow = dom.DivElement()
      ..className = 'af-text'
      ..style.marginTop = '8px'
      ..style.display = 'flex'
      ..style.alignItems = 'center'
      ..append(_icon)
      ..append(_text);

    _content..append(titleRow)..append(textRow);
    _container.append(_content);
    dom.document.body?.append(_container);

    if (dom.document.head?.querySelector('#af-toast-style') == null) {
      final css = dom.StyleElement()
        ..id = 'af-toast-style'
        ..innerHtml = '''
          div.af-title h2.af-heading {
            all: initial;
          }
          div.af-title div.af-button-close {
            all: initial;
          }
          div.af-text p.af-text {
            all: initial;
          }
        ''';
      dom.document.head?.children.add(css);
    }

    if (duration != null) {
      timer = Timer(duration, close);
    }

    _closeButton.onClick.first.then((_) => close());
    Future.delayed(const Duration(milliseconds: 10))
        .then((_) => evaluatePositions());
  }

  void close() async {
    timer?.cancel();
    _container.style.opacity = '0';
    await Future.delayed(const Duration(milliseconds: 300));
    _container.remove();
    evaluatePositions();
  }
}

void evaluatePositions() {
  final elements =
      dom.document.body?.querySelectorAll('div.af-toast-container');
  if (elements != null) {
    var bottom = _toastMargin;
    for (final element in elements) {
      element.style.bottom = '${bottom}px';
      bottom += _toastMargin + element.clientHeight;
    }
  }
}
